src/Java.Interop/Java.Interop/JavaPrimitiveArrays.cs

```diff
@@ -12,50 +12,44 @@
-				return true;
-			}
-			if (type == typeof (JavaArray<Int32>) || type == typeof (JavaPrimitiveArray<Int32>)) {
-				signature = GetCachedTypeSignature (ref __Int32TypeArraySignature, "I", arrayRank: 1, keyword: true);
-				return true;
-			}
-			if (type == typeof (JavaArray<Int64>) || type == typeof (JavaPrimitiveArray<Int64>)) {
-				signature = GetCachedTypeSignature (ref __Int64TypeArraySignature, "J", arrayRank: 1, keyword: true);
-				return true;
-			}
-			if (type == typeof (JavaArray<Single>) || type == typeof (JavaPrimitiveArray<Single>)) {
-				signature = GetCachedTypeSignature (ref __SingleTypeArraySignature, "F", arrayRank: 1, keyword: true);
-				return true;
-			}
-			if (type == typeof (JavaArray<Double>) || type == typeof (JavaPrimitiveArray<Double>)) {
-				signature = GetCachedTypeSignature (ref __DoubleTypeArraySignature, "D", arrayRank: 1, keyword: true);
-				return true;
-			}
-			return false;
+
+		partial class JniTypeManager {
+
+			readonly struct JniPrimitiveArrayInfo {
+				public  readonly    JniTypeSignature    JniTypeSignature;
+				public  readonly    Type                PrimitiveType;
+				public  readonly    Type[]              ArrayTypes;
+
+				public JniPrimitiveArrayInfo (string jniSimpleReference, Type primitiveType, params Type[] arrayTypes)
+				{
+					JniTypeSignature    = new JniTypeSignature (jniSimpleReference, arrayRank: 1, keyword: true);
+					PrimitiveType       = primitiveType;
+					ArrayTypes          = arrayTypes;
+				}
+			}
+
+			static readonly JniPrimitiveArrayInfo[] JniPrimitiveArrayTypes = new JniPrimitiveArrayInfo[]{
+				new ("Z", typeof (Boolean), typeof (Boolean[]), typeof (JavaArray<Boolean>), typeof (JavaPrimitiveArray<Boolean>), typeof (JavaBooleanArray)),
+				new ("B", typeof (SByte), typeof (SByte[]), typeof (JavaArray<SByte>), typeof (JavaPrimitiveArray<SByte>), typeof (JavaSByteArray)),
+				new ("C", typeof (Char), typeof (Char[]), typeof (JavaArray<Char>), typeof (JavaPrimitiveArray<Char>), typeof (JavaCharArray)),
+				new ("S", typeof (Int16), typeof (Int16[]), typeof (JavaArray<Int16>), typeof (JavaPrimitiveArray<Int16>), typeof (JavaInt16Array)),
+				new ("I", typeof (Int32), typeof (Int32[]), typeof (JavaArray<Int32>), typeof (JavaPrimitiveArray<Int32>), typeof (JavaInt32Array)),
+				new ("J", typeof (Int64), typeof (Int64[]), typeof (JavaArray<Int64>), typeof (JavaPrimitiveArray<Int64>), typeof (JavaInt64Array)),
+				new ("F", typeof (Single), typeof (Single[]), typeof (JavaArray<Single>), typeof (JavaPrimitiveArray<Single>), typeof (JavaSingleArray)),
+				new ("D", typeof (Double), typeof (Double[]), typeof (JavaArray<Double>), typeof (JavaPrimitiveArray<Double>), typeof (JavaDoubleArray)),
+			};
+
+			static bool GetBuiltInTypeArraySignature (Type type, ref JniTypeSignature signature)
+			{
+				foreach (var e in JniPrimitiveArrayTypes) {
+					if (Array.IndexOf (e.ArrayTypes, type) < 0)
+						continue;
```

Comment:
Probably not, and we're talking O(N**2) on *8 entries*, each containing *4* elements within `e.ArrayTypes`, or *32 entries* total.  O(N**2) on 32 entries should be negligible.

> Is there a way to rework the data to be a Dictionary<Type, JniPrimitiveArrayInfo> to make this lookup faster?

Yes, but as this is static data, it'll be constructed during app startup, *even for .NET Android* (unless we make it conditional, and please no?), so I'm trying to figure out how to *minimize* startup overhead.

My assumption (un-timed) is that creating an array of structs will be faster than populating a Dictionary, particularly when we're dealing with so few elements (8).

Reply:
So, about that init vs. lookup time…

<details>
<summary>Repro App</summary>
```csharp

using System.Diagnostics;

const int CreationCount = 1000000;
const int LookupCount   = 1000000;

var a = Stopwatch.StartNew ();
for (int i = 0; i < CreationCount; ++i) {
    CreateArray ();
}
a.Stop ();

var d = Stopwatch.StartNew ();
for (int i = 0; i < CreationCount; ++i) {
    CreateDict ();
}
d.Stop ();

Console.WriteLine ($"Array Creation: {a.ElapsedMilliseconds}ms");
Console.WriteLine ($" Dict Creation: {d.ElapsedMilliseconds}ms");

var av = CreateArray ();
a = Stopwatch.StartNew ();
for (int i = 0; i < LookupCount; ++i) {
    TryArrayLookup (av, typeof (int));
}
a.Stop ();

var bv = CreateDict ();
for (int i = 0; i < LookupCount; ++i) {
    TryDictLookup (bv, typeof (int));
}
d.Stop ();
Console.WriteLine ($"Array Lookup: {a.ElapsedMilliseconds}ms");
Console.WriteLine ($" Dict Lookup: {d.ElapsedMilliseconds}ms");

static JniPrimitiveArrayInfo[] CreateArray ()
{
	return new JniPrimitiveArrayInfo[]{
		new ("Z", typeof (Boolean), typeof (Boolean[]), typeof (JavaArray<Boolean>), typeof (JavaPrimitiveArray<Boolean>), typeof (JavaBooleanArray)),
		new ("B", typeof (SByte),   typeof (SByte[]), typeof (JavaArray<SByte>), typeof (JavaPrimitiveArray<SByte>), typeof (JavaSByteArray)),
		new ("C", typeof (Char),    typeof (Char[]), typeof (JavaArray<Char>), typeof (JavaPrimitiveArray<Char>), typeof (JavaCharArray)),
		new ("S", typeof (Int16),   typeof (Int16[]), typeof (JavaArray<Int16>), typeof (JavaPrimitiveArray<Int16>), typeof (JavaInt16Array)),
		new ("I", typeof (Int32),   typeof (Int32[]), typeof (JavaArray<Int32>), typeof (JavaPrimitiveArray<Int32>), typeof (JavaInt32Array)),
		new ("J", typeof (Int64),   typeof (Int64[]), typeof (JavaArray<Int64>), typeof (JavaPrimitiveArray<Int64>), typeof (JavaInt64Array)),
		new ("F", typeof (Single),  typeof (Single[]), typeof (JavaArray<Single>), typeof (JavaPrimitiveArray<Single>), typeof (JavaSingleArray)),
		new ("D", typeof (Double),  typeof (Double[]), typeof (JavaArray<Double>), typeof (JavaPrimitiveArray<Double>), typeof (JavaDoubleArray)),
    };
}

static bool TryArrayLookup (JniPrimitiveArrayInfo[] array, Type type)
{
	foreach (var e in array) {
		if (e.PrimitiveType != type)
			continue;
		return true;
	}
    return false;
}

static Dictionary<Type, JniPrimitiveArrayInfo> CreateDict ()
{
    return new Dictionary<Type, JniPrimitiveArrayInfo>  () {
		[typeof (Boolean)] = new ("Z", typeof (Boolean), typeof (Boolean[]), typeof (JavaArray<Boolean>), typeof (JavaPrimitiveArray<Boolean>), typeof (JavaBooleanArray)),
		[typeof (SByte)  ] = new ("B", typeof (SByte), typeof (SByte[]), typeof (JavaArray<SByte>), typeof (JavaPrimitiveArray<SByte>), typeof (JavaSByteArray)),
		[typeof (Char)   ] = new ("C", typeof (Char), typeof (Char[]), typeof (JavaArray<Char>), typeof (JavaPrimitiveArray<Char>), typeof (JavaCharArray)),
		[typeof (Int16)  ] = new ("S", typeof (Int16), typeof (Int16[]), typeof (JavaArray<Int16>), typeof (JavaPrimitiveArray<Int16>), typeof (JavaInt16Array)),
		[typeof (Int32)  ] = new ("I", typeof (Int32), typeof (Int32[]), typeof (JavaArray<Int32>), typeof (JavaPrimitiveArray<Int32>), typeof (JavaInt32Array)),
		[typeof (Int64)  ] = new ("J", typeof (Int64), typeof (Int64[]), typeof (JavaArray<Int64>), typeof (JavaPrimitiveArray<Int64>), typeof (JavaInt64Array)),
		[typeof (Single) ] = new ("F", typeof (Single), typeof (Single[]), typeof (JavaArray<Single>), typeof (JavaPrimitiveArray<Single>), typeof (JavaSingleArray)),
		[typeof (Double) ] = new ("D", typeof (Double), typeof (Double[]), typeof (JavaArray<Double>), typeof (JavaPrimitiveArray<Double>), typeof (JavaDoubleArray)),
    };
}

static bool TryDictLookup (Dictionary<Type, JniPrimitiveArrayInfo> dict, Type type)
{
    if (dict.TryGetValue (type, out var _))
        return true;
    return false;
}

class JavaArray<T> {}
class JavaPrimitiveArray<T> : JavaArray<T> {}
class JavaBooleanArray : JavaPrimitiveArray<Boolean> {}
class JavaSByteArray : JavaPrimitiveArray<SByte> {}
class JavaCharArray : JavaPrimitiveArray<Char> {}
class JavaInt16Array : JavaPrimitiveArray<Int16> {}
class JavaInt32Array : JavaPrimitiveArray<Int32> {}
class JavaInt64Array : JavaPrimitiveArray<Int64> {}
class JavaSingleArray : JavaPrimitiveArray<Single> {}
class JavaDoubleArray : JavaPrimitiveArray<Double> {}

readonly struct JniPrimitiveArrayInfo {
	public  readonly    JniTypeSignature    JniTypeSignature;
	public  readonly    Type                PrimitiveType;
	public  readonly    Type[]              ArrayTypes;
	public JniPrimitiveArrayInfo (string jniSimpleReference, Type primitiveType, params Type[] arrayTypes)
	{
		JniTypeSignature    = new JniTypeSignature (jniSimpleReference, arrayRank: 1, keyword: true);
		PrimitiveType       = primitiveType;
		ArrayTypes          = arrayTypes;
	}
}

struct JniTypeSignature {
    public string SimpleName;
    public int ArrayRank;
    public bool Keyword;
    public JniTypeSignature (string simpleName, int arrayRank, bool keyword)
    {
        SimpleName  = simpleName;
        ArrayRank   = arrayRank;
        Keyword     = keyword;
    }
}
```
</details>

Results:

```sh
% dotnet run
Array Creation: 646ms
 Dict Creation: 1131ms
Array Lookup: 29ms
 Dict Lookup: 1131ms

% dotnet run -c Release
Array Creation: 389ms
 Dict Creation: 770ms
Array Lookup: 21ms
 Dict Lookup: 770ms
```

As expected, array lookup is faster -- half the time! -- and lookup is also much faster with arrays.

That said, this investigation did present a bug in `GetBuiltInTypeArraySignature()`: it shouldn't have been looking at `.ArrayTypes`, but instead `.PrimitiveType`, because the callsite in `GetTypeSignature(Type)` "unwraps" arrays, which kinda makes that particular codepath unusable, as in `GetBuiltInTypeArraySignature()` should just be removed entirely. :thinking:

Reply:
Apparently `<details/>` doesn't work like i thought, so here's the execution + summary:

Results:

```sh
% dotnet run
Array Creation: 646ms
 Dict Creation: 1131ms
Array Lookup: 29ms
 Dict Lookup: 1131ms

% dotnet run -c Release
Array Creation: 389ms
 Dict Creation: 770ms
Array Lookup: 21ms
 Dict Lookup: 770ms
```

As expected, array lookup is faster -- half the time! -- and lookup is also much faster with arrays.

That said, this investigation did present a bug in `GetBuiltInTypeArraySignature()`: it shouldn't have been looking at `.ArrayTypes`, but instead `.PrimitiveType`, because the callsite in `GetTypeSignature(Type)` "unwraps" arrays, which kinda makes that particular codepath unusable, as in `GetBuiltInTypeArraySignature()` should just be removed entirely. :thinking:

Reply:
…and `GetBuiltInTypeArraySignature()` is required; trying to remove it breaks unit tests.  Which means my "repro app" isn't accurate on the lookup side.  After updating it, arrays are still faster:

```
% dotnet run -c Release
Array Creation: 392ms
 Dict Creation: 753ms
Array Lookup: 156ms
 Dict Lookup: 753ms
```

Reply:
Seeing the manual work involved to test, should we have a BenchmarkDotNet project in this repo?

It seems like it would be useful to commit your benchmark above, and we'd have the option to run it later. BDN also does appropriate warmup, runs out of process, does some math, that might generally make it better than a manual benchmark.

Reply:
Yes, we should probably have one, and it should probably be in `tests/Java.Interop-PerformanceTests` (and/or I could "just" add the above app to `Java.Interop-PerformanceTests`…).